### ISSUE TITLE
HLDA over-branches: ~80x too many nodes vs tomotopy at identical hyperparameters (faithfulness)

### ISSUE THREAD

**Issue by nealcaren** [nealcaren/topica#615]
## Summary

topica's HLDA builds a **dramatically larger topic tree than the reference (tomotopy) at identical hyperparameters**, and the divergence appears in the **first path-resampling sweep**. This is both a faithfulness problem and the root cause of HLDA's poor scaling (#611): runtime is ~proportional to tree size, so an ~80x node blowup is ~80x of the ~300x speed gap.

## Evidence (2000-doc 20NG, depth 3, gamma=0.1, alpha=0.1, eta/beta=0.01, seed 1)

Node count vs iteration:

| iters | topica nodes | tomotopy live_k |
|-------|--------------|-----------------|
| 1     | **389**      | **7**           |
| 3     | 494          | 7               |
| 5     | 506          | 7               |
| 10    | 535          | 7               |
| 20    | 562          | 7               |

tomotopy is at 7 nodes after one sweep and never moves; topica is at 389 after one sweep and grows to ~562. **Same nested-CRP equations, same level sampler, same defaults** — verified against tomotopy's `HLDAModel.hpp` (its level sampler is `(numByTopic[l] + alpha[l]) * (nw+eta)/(n+V*eta)`, symmetric Dirichlet, same eta/gamma; no depth bias, no stick-breaking).

## Decomposition (each measured, not assumed)

- **Not gamma:** matched gamma=0.1 -> still 542 (topica) vs 7 (tomotopy).
- **Not the level prior:** asymmetric alpha / GEM (#613) only *relocate* the bloat — root-heavy GEM m=0.9 cut leaf occupancy 41%->15% but level-1 nodes exploded 146->246 (total ~592).
- **Not alpha optimization:** disabling tomotopy's optimizer -> still 7.
- **Sensitive to beta in topica only:** topica beta 0.01/0.1/0.5/1.0 -> 542/95/27/13 nodes; tomotopy is flat at 7 for eta 0.01/0.1/1.0.

## Mechanism (hypothesis)

The collapsed Dirichlet-multinomial gives a **fresh node** a denominator advantage of ~`M * ln((n+V*eta)/(V*eta))` nats (M = a document's tokens at that level) that exceeds the nCRP sharing prior (~`ln(n_child_customers/gamma)`). With topica's ~40% leaf occupancy, M~=10-30, so nearly every non-matching document branches. The nCRP *prior* only wants ~3 branches (`gamma/(n+gamma)`); topica overrides it ~50x. The reference, with identical specs, does not — so topica's path-move candidate scoring (most likely the fresh-node `empty_level_log_marginal` term, or a normalization the reference has) is the suspected defect.

## Next

Line-level diagnosis of `sample_path` candidate scoring (prior vs likelihood, share vs new-branch) for a single document, term-by-term against the reference. Practical workaround today: `beta>=0.5` gives a compact, fast tree, but changes topic granularity.

Related: #611 (scaling), #613 (level priors), and the merged #614 (per-node memoization, orthogonal 1.6x).

**Comment by nealcaren** [comment]
## Line-level diagnosis

Instrumented `sample_path`. Key findings on the same 2000-doc / depth-3 / gamma=0.1 / eta=0.01 setup:

**It is the branching decision, not occupancy or init size.**
- topica post-init (iters=0): **11 nodes**. tomotopy pre-train: ~0. So neither starts bushy.
- First sweep: topica 11 -> **389**; tomotopy ~0 -> **7 and stable**. The entire divergence is sweep 1's path move.
- Per-level token occupancy is nearly identical: topica root/L1/leaf = 34/24/41%, tomotopy = 49/21/31%. tomotopy packs leaf tokens into **4 big leaves** (~13k tokens each); topica spreads them across **395 tiny ones**.

**The branch decision scales with document length.** Dumping best "share an existing full path" vs best "new branch" score for the first resamples:

```
docM=44   best_share=-398.73  best_new=-416.72  -> share wins (by 18)
docM=44   best_share=-389.05  best_new=-412.06  -> share wins
docM=195  best_share=-1678.01 best_new=-1660.70 -> NEW WINS (by 17), new_from=1
docM=11   best_share=-81.70   best_new=-100.16  -> share wins
docM=304  best_share=-2826.17 best_new=-2842.54 -> share wins (by 16)
```

Short docs share decisively; **long/token-heavy docs branch**. This matches the mechanism: the fresh-node Dirichlet-multinomial denominator bonus ~`M_level * ln((n+V*eta)/(V*eta))` scales with a document's tokens-per-level and overcomes the nCRP sharing prior (~`ln(n_customers/gamma)`).

**Why the reference avoids it (hypothesis).** Same formula, so the difference is the *state the formula sees*: topica's random-level init seeds leaves with incoherent word mixes, so no existing leaf matches a document's words (low share-numerator) -> long docs branch -> more incoherent small leaves -> snowball that never collapses (11->389->562 over 20 iters). tomotopy accumulates coherent leaves that documents match, so the share-numerator wins and it stabilizes at 7.

## Candidate fixes to try
1. **Coherent initialization** (the leading suspect): seed all documents onto a single shared path / assign tokens to the root initially and let the tree grow, instead of random per-token levels + greedy CRP paths. Test whether topica then stabilizes at a tomotopy-like node count.
2. Re-derive/verify the new-branch (`empty_level_log_marginal`) term against tomotopy's `lgamma_subt` path likelihood for an identical (node-counts, doc) pair to rule out a scoring discrepancy.

Next step is (1): a minimal-init experiment.

**Comment by nealcaren** [comment]
## Update: initialization is NOT the cause

Tested a coherent init (all tokens start at the root, leaves empty) via a temporary flag: **still 550 nodes** (vs 542 for the random-level init). So the snowball is not seeded by init — the path move over-branches regardless of the starting state. The level sampler pulls ~30% of tokens down to the leaves (from all-root init), and the path move then branches on them.

## Refined mechanism

The tension is on **shared nodes**: an intermediate/leaf node shared by many documents accumulates a large token count `n`. The collapsed Dirichlet-multinomial penalizes adding a document's `M` tokens to a large-`n` node by `~M*ln((n+V*eta)/(V*eta))`. For a document whose words do not match that node's distribution (distinctive-vocabulary docs, i.e. most of 20NG), a **fresh node wins** because it has no such penalty. So topica branches for every non-matching document.

tomotopy has the **identical** marginal formula (`lgamma_subt` over the same `numByTopic`/`numByTopicWord` + eta counts) and the same shared-node structure, yet stabilizes at 7 nodes. So there is a real computational difference in their path-sampling routine that I have not yet isolated from the header snippets — the next step is a term-by-term comparison of topica's per-candidate score vs tomotopy's for an identical (tree-state, document) pair, reading the full `sampleTopics`/`getZLikelihoods` path in their source (or instrumenting both). Ruling out: gamma, level prior, alpha-optimization, occupancy, and init (all tested).

**Comment by nealcaren** [comment]
## RESOLVED: term-by-term identical; the cause is sampler metastability, not a wrong formula

Read tomotopy's actual C++ (`src/TopicModel/HLDAModel.hpp`) and compared every term of the path-resampling score to topica's `sample_path`:

- **nCRP prior** — tomotopy `updateNodeLikelihood`: `log(child.numCustomers/(parent.numCustomers+gamma))` per descent, `log(gamma/(node.numCustomers+gamma))` for a new child. topica: identical.
- **Existing-node word marginal** — tomotopy `calcWordLikelihood`: `sum_w lgamma_subt(numByTopicWord+eta, cnt)` minus `lgamma_subt(numByTopic+realV*eta, M)`. topica `level_log_marginal`: identical.
- **New-node marginal** (`newTopicWeights`): `sum_w[lgamma(eta+m)-lgamma(eta)] - [lgamma(M+V*eta)-lgamma(V*eta)]`. topica `empty_level_log_marginal`: identical.
- Root-likelihood elision (tomotopy) is a constant across candidates -> no effect on sampling.

**The formulas are the same, so the samplers share a stationary distribution.** The divergence is convergence/mixing:

| iters (500 docs) | topica nodes | tomotopy live_k |
|------------------|--------------|-----------------|
| 10  | 318 | 6 |
| 100 | 290 | 6 |
| 400 | 262 | 6 |

topica is **trapped in a bushy metastable state** — 400 iterations only nudge it 318->262, nowhere near 6. Mechanism: single-document path moves cannot *merge* many small mutually-incoherent leaves (no document finds a clearly better leaf to join), so the configuration is frozen. Seeding a compact 3-node tree does not help either, because that single leaf is the global average (incoherent), so branching is locally correct and it re-explodes. tomotopy's incremental, prior-sampled initialization lands in the compact coherent-leaf basin and stays.

## Fixes (ranked)
1. **Split-merge / restaurant-level moves** (principled): add an MH merge move so the sampler can escape the bushy basin. Standard remedy for nCRP/DP-mixture metastability. Most work, real fix.
2. **Match tomotopy's initialization**: incremental, prior-based path assignment to start in the compact basin. Cheaper, but may not hold if it doesn't also seed coherent leaves (the single-path test is a caution).
3. **Practical workaround now**: `beta >= 0.5` gives a compact, fast, tomotopy-like tree (13-27 nodes) — the peaked default `beta=0.01` is what makes small leaves attractive. Document this.

Recommend documenting (3) immediately and pursuing (1) for a real fix. This is a faithfulness/quality issue independent of the merged perf work (#614).

**Comment by nealcaren** [comment]
## Correction: this is NOT a bug. The bushy tree is the genuine posterior mode.

Building the split-merge move forced the decisive measurement I should have made first: the **collapsed joint log-probability** log P(w, c, z) of the tree. Results on the 500-doc 20NG (depth 3, gamma=0.1):

**The sampler finds the posterior mode at each beta:**

| beta  | nodes | log_joint (at that beta) |
|-------|-------|--------------------------|
| 0.01  | 300   | -422809 |
| 0.1   | 81    | -424273 |
| 1.0   | 10    | -433623 |

**Force-collapsing the beta=0.01 tree LOWERS the joint:** merging down to ~165 nodes drops log_joint to ~-450000 to -466000 (vs -422809 for the natural 300-node fit). A correctly-implemented split-merge move confirms this independently — it *rejects* merges, because every merge lowers the posterior.

### What this means
Under the sharp default beta=0.01, the fine-grained (many-node) tree genuinely has the **highest collapsed posterior**: a sharp topic-word Dirichlet rewards many pure/coherent nodes, and that reward outweighs the nCRP prior's penalty for extra nodes. topica's collapsed Gibbs **correctly locates this mode**. tomotopy's ~7 nodes at the same hyperparameters is a **lower-posterior** configuration — tomotopy under-fits / under-explores; it does not split into the finer tree the model actually prefers.

**The earlier "over-branching / metastability" diagnosis in this issue was wrong.** There is no metastable trap: the plateau at ~262-300 nodes is the posterior mode, not a barrier. The node count is a **beta-controlled modeling property**, not a bug — larger beta gives a compact tree (beta=1.0 -> 10 nodes) and is the right knob for a user who wants coarse topics.

### Disposition
- topica HLDA is **faithful and correct** (arguably fits *better* than the reference at sharp beta). No fix needed.
- The split-merge move (branch `feat/hlda-split-merge-615`) is correctly implemented and detailed-balance-valid, but it does **not** collapse the tree — because the posterior does not favor a collapsed tree. It solves a non-problem and will not ship as a #615 fix.
- **Actionable:** document that HLDA's tree granularity is governed by `beta` (small = many fine topics, large = few coarse topics), so users can pick the resolution they want. Recommend closing this issue as "not a bug".

**Comment by nealcaren** [comment]
Closing as not-a-bug. Root cause determined: HLDA node count is the beta-controlled collapsed-posterior mode (small beta -> many fine nodes, large beta -> few coarse), verified by joint log-probability. topica correctly follows the posterior; the reference under-fits at sharp beta. Documented in docs/guides/models.md (commit 6b84ad4). The split-merge move (branch feat/hlda-split-merge-615) is correct but unnecessary and will not ship.
